File: CryptoApp/Shared Components/SkeletonView/CoinCell/CoinCellSkeleton.m
//
//  CoinCellSkeleton.m
//  CryptoApp
//
//  Created by Jansen Castillo
//

#import "CoinCellSkeleton.h"
#import "SkeletonView.h"

@interface CoinCellSkeleton ()

@property (nonatomic, strong) NSArray<SkeletonView *> *skeletonViews;

// Layout components to match CoinCell
@property (nonatomic, strong) SkeletonView *rankSkeleton;
@property (nonatomic, strong) SkeletonView *imageSkeleton;
@property (nonatomic, strong) SkeletonView *nameSkeleton;
@property (nonatomic, strong) SkeletonView *marketSkeleton;
@property (nonatomic, strong) SkeletonView *priceSkeleton;
@property (nonatomic, strong) SkeletonView *sparklineSkeleton;
@property (nonatomic, strong) SkeletonView *percentSkeleton;

// Stack views to match CoinCell layout
@property (nonatomic, strong) UIStackView *nameStack;
@property (nonatomic, strong) UIStackView *leftStack;
@property (nonatomic, strong) UIStackView *sparklineAndPercentStack;
@property (nonatomic, strong) UIStackView *rightStack;
@property (nonatomic, strong) UIStackView *mainStack;

@end

@implementation CoinCellSkeleton

// MARK: - Initialization

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
        [self collectSkeletonViews];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setupUI];
        [self collectSkeletonViews];
    }
    return self;
}

// MARK: - Setup

- (void)setupUI {
    [self createSkeletonViews];
    [self createStackViews];
    [self layoutStackViews];
    [self applyConstraints];
}

- (void)createSkeletonViews {
    self.rankSkeleton = [SkeletonView textSkeletonWithWidth:20 height:12];
    self.imageSkeleton = [SkeletonView circleSkeletonWithDiameter:32];
    self.nameSkeleton = [SkeletonView textSkeletonWithWidth:80 height:14];
    self.marketSkeleton = [SkeletonView textSkeletonWithWidth:60 height:12];
    self.priceSkeleton = [SkeletonView textSkeletonWithWidth:70 height:14];
    self.sparklineSkeleton = [SkeletonView rectangleSkeletonWithWidth:60 height:20 cornerRadius:2];
    self.percentSkeleton = [SkeletonView textSkeletonWithWidth:50 height:12];
}

- (void)createStackViews {
    // Name stack (vertical)
    self.nameStack = [[UIStackView alloc] initWithArrangedSubviews:@[self.nameSkeleton, self.marketSkeleton]];
    self.nameStack.axis = UILayoutConstraintAxisVertical;
    self.nameStack.spacing = 2;
    self.nameStack.alignment = UIStackViewAlignmentLeading;
    
    // Left stack (horizontal)
    self.leftStack = [[UIStackView alloc] initWithArrangedSubviews:@[self.rankSkeleton, self.imageSkeleton, self.nameStack]];
    self.leftStack.axis = UILayoutConstraintAxisHorizontal;
    self.leftStack.alignment = UIStackViewAlignmentCenter;
    self.leftStack.spacing = 12;
    self.leftStack.distribution = UIStackViewDistributionFill;
    
    // Sparkline and percent stack (vertical)
    self.sparklineAndPercentStack = [[UIStackView alloc] initWithArrangedSubviews:@[self.sparklineSkeleton, self.percentSkeleton]];
    self.sparklineAndPercentStack.axis = UILayoutConstraintAxisVertical;
    self.sparklineAndPercentStack.spacing = 4;
    self.sparklineAndPercentStack.alignment = UIStackViewAlignmentCenter;
    
    // Right stack (horizontal)
    self.rightStack = [[UIStackView alloc] initWithArrangedSubviews:@[self.priceSkeleton, self.sparklineAndPercentStack]];
    self.rightStack.axis = UILayoutConstraintAxisHorizontal;
    self.rightStack.alignment = UIStackViewAlignmentCenter;
    self.rightStack.spacing = 16;
    self.rightStack.distribution = UIStackViewDistributionFill;
    
    // Main stack (horizontal)
    self.mainStack = [[UIStackView alloc] initWithArrangedSubviews:@[self.leftStack, self.rightStack]];
    self.mainStack.axis = UILayoutConstraintAxisHorizontal;
    self.mainStack.alignment = UIStackViewAlignmentCenter;
    self.mainStack.distribution = UIStackViewDistributionFillProportionally;
    self.mainStack.spacing = 16;
    self.mainStack.translatesAutoresizingMaskIntoConstraints = NO;
}

- (void)layoutStackViews {
    [self.contentView addSubview:self.mainStack];
}

- (void)applyConstraints {
    // Match CoinCell constraints exactly
    [NSLayoutConstraint activateConstraints:@[
        [self.mainStack.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:12],
        [self.mainStack.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:16],
        [self.mainStack.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor constant:-16],
        [self.mainStack.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor constant:-12],
        
        // Fixed sizes to match CoinCell
        [self.rankSkeleton.widthAnchor constraintGreaterThanOrEqualToConstant:20],
        [self.nameSkeleton.widthAnchor constraintGreaterThanOrEqualToConstant:80],
        [self.priceSkeleton.widthAnchor constraintGreaterThanOrEqualToConstant:70],
        [self.sparklineAndPercentStack.widthAnchor constraintGreaterThanOrEqualToConstant:60]
    ]];
    
    // Set content hugging priorities to match CoinCell
    [self.rankSkeleton setContentHuggingPriority:UILayoutPriorityDefaultHigh forAxis:UILayoutConstraintAxisHorizontal];
    [self.imageSkeleton setContentHuggingPriority:UILayoutPriorityDefaultHigh forAxis:UILayoutConstraintAxisHorizontal];
    [self.sparklineSkeleton setContentHuggingPriority:UILayoutPriorityDefaultHigh forAxis:UILayoutConstraintAxisHorizontal];
    [self.sparklineAndPercentStack setContentHuggingPriority:UILayoutPriorityDefaultHigh forAxis:UILayoutConstraintAxisHorizontal];
}

- (void)collectSkeletonViews {
    self.skeletonViews = @[
        self.rankSkeleton,
        self.imageSkeleton,
        self.nameSkeleton,
        self.marketSkeleton,
        self.priceSkeleton,
        self.sparklineSkeleton,
        self.percentSkeleton
    ];
}

// MARK: - Public Methods

- (void)startShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView startShimmering];
    }
}

- (void)stopShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView stopShimmering];
    }
}

// MARK: - Reuse

- (void)prepareForReuse {
    [super prepareForReuse];
    [self stopShimmering];
}

+ (NSString *)reuseID {
    return @"CoinCellSkeleton";
}

@end

File: CryptoApp/Shared Components/SkeletonView/CoinCellSkeleton.swift
import UIKit

final class CoinCellSkeleton: UICollectionViewCell {
    
    // MARK: - Properties
    
    private var skeletonViews: [SkeletonView] = []
    
    // Layout components to match CoinCell
    private let rankSkeleton = SkeletonView.textSkeleton(width: 20, height: 12)
    private let imageSkeleton = SkeletonView.circleSkeleton(diameter: 32)
    private let nameSkeleton = SkeletonView.textSkeleton(width: 80, height: 14)
    private let marketSkeleton = SkeletonView.textSkeleton(width: 60, height: 12)
    private let priceSkeleton = SkeletonView.textSkeleton(width: 70, height: 14)
    private let sparklineSkeleton = SkeletonView.rectangleSkeleton(width: 60, height: 20, cornerRadius: 2)
    private let percentSkeleton = SkeletonView.textSkeleton(width: 50, height: 12)
    
    // Stack views to match CoinCell layout
    private lazy var nameStack: UIStackView = {
        let stack = UIStackView(arrangedSubviews: [nameSkeleton, marketSkeleton])
        stack.axis = .vertical
        stack.spacing = 2
        stack.alignment = .leading
        return stack
    }()
    
    private lazy var leftStack: UIStackView = {
        let stack = UIStackView(arrangedSubviews: [rankSkeleton, imageSkeleton, nameStack])
        stack.axis = .horizontal
        stack.alignment = .center
        stack.spacing = 12
        stack.distribution = .fill
        return stack
    }()
    
    private lazy var sparklineAndPercentStack: UIStackView = {
        let stack = UIStackView(arrangedSubviews: [sparklineSkeleton, percentSkeleton])
        stack.axis = .vertical
        stack.spacing = 4
        stack.alignment = .center
        return stack
    }()
    
    private lazy var rightStack: UIStackView = {
        let stack = UIStackView(arrangedSubviews: [priceSkeleton, sparklineAndPercentStack])
        stack.axis = .horizontal
        stack.alignment = .center
        stack.spacing = 16
        stack.distribution = .fill
        return stack
    }()
    
    private lazy var mainStack: UIStackView = {
        let stack = UIStackView(arrangedSubviews: [leftStack, rightStack])
        stack.axis = .horizontal
        stack.alignment = .center
        stack.distribution = .fillProportionally
        stack.spacing = 16
        stack.translatesAutoresizingMaskIntoConstraints = false
        return stack
    }()
    
    // MARK: - Initialization
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        collectSkeletonViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupUI()
        collectSkeletonViews()
    }
    
    // MARK: - Setup
    
    private func setupUI() {
        contentView.addSubview(mainStack)
        
        // Match CoinCell constraints exactly
        NSLayoutConstraint.activate([
            mainStack.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 12),
            mainStack.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16),
            mainStack.trailingAnchor.constraint(equalTo: contentView.trailingAnchor, constant: -16),
            mainStack.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -12),
            
            // Fixed sizes to match CoinCell
            rankSkeleton.widthAnchor.constraint(greaterThanOrEqualToConstant: 20),
            nameSkeleton.widthAnchor.constraint(greaterThanOrEqualToConstant: 80),
            priceSkeleton.widthAnchor.constraint(greaterThanOrEqualToConstant: 70),
            sparklineAndPercentStack.widthAnchor.constraint(greaterThanOrEqualToConstant: 60)
        ])
        
        // Set content hugging priorities to match CoinCell
        rankSkeleton.setContentHuggingPriority(.defaultHigh, for: .horizontal)
        imageSkeleton.setContentHuggingPriority(.defaultHigh, for: .horizontal)
        sparklineSkeleton.setContentHuggingPriority(.defaultHigh, for: .horizontal)
        sparklineAndPercentStack.setContentHuggingPriority(.defaultHigh, for: .horizontal)
    }
    
    private func collectSkeletonViews() {
        skeletonViews = [
            rankSkeleton,
            imageSkeleton,
            nameSkeleton,
            marketSkeleton,
            priceSkeleton,
            sparklineSkeleton,
            percentSkeleton
        ]
    }
    
    // MARK: - Public Methods
    
    func startShimmering() {
        skeletonViews.forEach { $0.startShimmering() }
    }
    
    func stopShimmering() {
        skeletonViews.forEach { $0.stopShimmering() }
    }
    
    // MARK: - Reuse
    
    override func prepareForReuse() {
        super.prepareForReuse()
        stopShimmering()
    }
    
    static var reuseID: String {
        return "CoinCellSkeleton"
    }
} 